boto/endpoints.py

```diff
@@ -0,0 +1,415 @@
+        # A mapping can be provided that handles the mapping between
+        # "service names" and endpoint prefixes.
+        'awslambda': 'lambda',
+        'cloudwatch': 'monitoring',
+        'ses': 'email',
+        'ec2containerservice': 'ecs',
+        'configservice': 'config',
+    }
+
+    def __init__(self, endpoint_data, boto_endpoint_data=None,
+                 service_rename_map=None):
+        """
+        :type endpoint_data: dict
+        :param endpoint_data: Regions and endpoints data in the same format
+            as is used by botocore / boto3.
+
+        :type boto_endpoint_data: dict
+        :param boto_endpoint_data: Regions and endpoints data in the legacy
+            boto2 format. This data takes precedence over any data found in
+            `endpoint_data`.
+
+        :type service_rename_map: dict
+        :param service_rename_map: A mapping of boto2 service name to
+            endpoint prefix.
+        """
+        super(BotoEndpointResolver, self).__init__(endpoint_data)
+        self._boto_endpoint_data = boto_endpoint_data
+        if service_rename_map is None:
+            service_rename_map = self.SERVICE_RENAMES
+        # Mapping of boto2 service name to endpoint prefix
+        self._endpoint_prefix_map = service_rename_map
+        # Mapping of endpoint prefix to boto2 service name
+        self._service_name_map = dict(
+            (v, k) for k, v in service_rename_map.items())
+
+    def get_available_endpoints(self, service_name, partition_name='aws',
+                                allow_non_regional=False):
+        endpoint_prefix = self._endpoint_prefix(service_name)
+        endpoints = super(BotoEndpointResolver, self).get_available_endpoints(
+            endpoint_prefix, partition_name, allow_non_regional)
+        boto_endpoints = self._get_available_boto_endpoints(service_name)
+        return list(set(endpoints + boto_endpoints))
+
+    def get_all_available_regions(self, service_name):
+        """Retrieve every region across partitions for a service."""
+        regions = set()
+        endpoint_prefix = self._endpoint_prefix(service_name)
+
+        # Get every region for every partition in the new endpoint format
+        for partition_name in self.get_available_partitions():
+            if self.is_global_service(service_name, partition_name):
+                # Global services are available in every region in the
+                # partition in which they are considered global.
+                partition = self._get_partition_data(partition_name)
+                regions.update(partition['regions'].keys())
+            else:
+                regions.update(
+                    super(BotoEndpointResolver, self).get_available_endpoints(
+                        endpoint_prefix, partition_name
+                    )
+                )
+
+        # boto2 endpoint format has no partitions, so load them all.
+        boto_endpoints = self._get_available_boto_endpoints(service_name)
```

Comment:
> I don't think it makes sense to leak the concerns of resolving an endpoint to a builder. I could understand moving it to a different class, but then you'll have to re-implement the prioritization logic everywhere or create a third class that stitches the two together

I want to make sure we're on the same page in terms of what's being proposed.  I think the main disconnect here is I (and I believe Kyle as well) are suggesting to *not* expose a "BotoEndpointResolver" as a first class concept.  I propose that the class should *should not even* exist.  

Why?

Consider `regioninfo.load_regions()`.  The only thing that matters to callers of that function is that we return a dictionary of `{service: {region: endpoint}}`.  How we construct that is an implementation detail, as long as we return that dict we can create it however we want.

So what's the least invasive/simplest way to make this change?

Right now the code in the `develop` branch is this:

```python
def load_regions():
    endpoints = load_endpoint_json(boto.ENDPOINTS_PATH)
    additional_path = None
    if os.environ.get('BOTO_ENDPOINTS'):
        additional_path = os.environ['BOTO_ENDPOINTS']
    elif boto.config.get('Boto', 'endpoints_path'):
        additional_path = boto.config.get('Boto', 'endpoints_path')
    if additional_path:
        additional = load_endpoint_json(additional_path)
        endpoints = merge_endpoints(endpoints, additional)
    return endpoints
```

The first line of that function loads a JSON file that's already in the exact structure we need.  So the simplest way to change this would be to change that first line to:

```
def load_regions():
    endpoints = construct_legacy_format_using_botocore_endpoints_format()
    ... everything else is the same ...
```

Where the job of this new function is to take botocore's endpoints.json and transform them into the legacy format.  It has a single job.  It does know about `BOTO_ENDPOINTS` or overrides or anything.  Take botocore endpoints -> produce legacy endpoint format.  That's it.

Let's say we stopped there and forgot about doing endpoint heuristics.

This is a pretty low risk change.  In fact we could pretty definitively know that nothing will break.  Ensure that `load_regions()` in the develop branch and `load_regions()` in this PR return the exact same results.  If that's the case we're fine.

It's also really simple to review.  With the exception of one line of code changed in `load_regions()` everything else is entirely brand new code that's not touching anything else in boto.  And the new code is purely about transforming one format to another.